### ISSUE TITLE
Bazel mode doesn't automatically highlight all .bazel files.

### ISSUE THREAD

**Issue by yesudeep** [bazelbuild/emacs-bazel-mode#309]
Some build files that are written for third party libraries don't get bazel mode enabled. I needed to add:

```
(add-to-list 'auto-mode-alist
             '("\\.bazel" . bazel-build-mode))
```

to my configuration to enable bazel mode for those files.

Would it be possible to add this directly to the mode?

**Comment by yesudeep** [comment]
Hi @phst I've updated my PR to use a predicate to determine whether 
the visited buffer file is a WORKSPACE or a BUILD file. Using that with magic-mode-alist
instead of auto-mode-alist appears to use `bazel-workspace-mode` for workspace files
and `bazel-build-mode` for build files.

Could you please take a look at the PR? 
https://github.com/bazelbuild/emacs-bazel-mode/pull/310

**Comment by phst** [comment]
Thanks for the pointer to https://github.com/mathetake/bazel-wasmtime-c-api/blob/03dea33441a2d88fd494556da0f2e690b7cd5f14/bazel/cargo/crates.bzl. From what I can see, all the files in https://github.com/mathetake/bazel-wasmtime-c-api/tree/03dea33441a2d88fd494556da0f2e690b7cd5f14/bazel/cargo/remote are of the form `BUILD.*.bazel`, so wouldn't it be enough to add a pattern that matches that form?
I'm trying to find a balance here between auto-detecting files as correctly as possible and avoiding too much complexity or "magic."
Are there many examples of files that can't be matched by a set of specific regexp pattern? I unfortunately don't have a good overview what typical conventions exist.

**Comment by yesudeep** [comment]
I'm not sure but suggest the simplest way to tackle this would be to associate it as follows while striking a balance between bazel-workspace-mode and bazel-build-mode:

1. BUILD, *.bazel (excluding WORKSPACE.bazel) =>  bazel-build-mode
2. WORKSPACE, WORKSPACE.bazel => bazel-workspace-mode

Re `BUILD.*.bazel`: I'm not sure that would work, because one can find examples of people using other formats as well:

Here's an example: https://github.com/ttiurani/bazel-prost-build-bug/blob/fe4acaeccfe584b1f9f18ce35b9fe689198ab844/bazel/third_party/cargo/crates.bzl#L38

`rules_foreign_cc` provides rules to integrate with different build systems (cmake, autotools, etc.) and if tomorrow someone decides to integrate meson, one will have to deal with the ambiguity between meson's `*.build` file and bazel's `*.BUILD` file (especially on case insensitive file systems). I'd say it's far easier to strike a balance by having Bazel (and tools around it) formally use `.bazel` and `.bzl` file extensions. It would avoid a lot of ambiguity and guess work.

**Comment by phst** [comment]
> I'd say it's far easier to strike a balance by having Bazel (and tools around it) formally use .bazel and .bzl file extensions.

I don't disagree, but I guess it would be better to discuss that in https://github.com/bazelbuild/bazel-website and have bazel.el follow conventions established elsewhere.

> Re BUILD.*.bazel: I'm not sure that would work, because one can find examples of people using other filename formats as well:

Yeah, and we should definitely add patterns to match these conventions, e.g. something like `foo.BUILD`, `BUILD.foo`, each with an optional `.bazel` suffix, and similar for workspaces.

> What do you think?

As said before, I think the conventions and recommendations should be documented in the main Bazel documentation. Some additions look unambiguous enough, e.g. `foo.BUILD.bazel` or `BUILD.foo.bazel`, so feel free to add them right away.

> Build files for remote workspaces defined within the current workspace: *.bazel (e.g. com_google_fruit.bazel) while excluding BUILD.bazel and WORKSPACE.bazel because those would imply the current workspace. It's also possible to consider *.BUILD.bazel, but I'm not sure whether that would be redundant (are there any other types of .bazel file?).

I'd exclude `*.bazel` filenames that don't contain `BUILD` anywhere, they might also contain remote workspace definitions. Or is there some widely-used convention that these files unambiguously refer to BUILD files? At least the documented examples I found (https://docs.bazel.build/versions/main/repo/http.html#http_archive, https://docs.bazel.build/versions/main/external.html#non-bazel-projects, https://docs.bazel.build/versions/4.2.1/be/workspace.html#new_local_repository) all used either `foo.BUILD` or `BUILD.foo`, but not `foo.bazel`.

To summarize, my suggestion would be to detect `BUILD`, `foo.BUILD`, `BUILD.foo` (each with an optional `.bazel` extension) as BUILD files, and the same with WORKSPACE instead of BUILD.

The conflict with Meson and other build systems is somewhat annoying, unfortunately, but I think it's unavoidable since the "filename pattern space" is too limited.
